src/subcommand/server/cache.rs

```diff
@@ -1,4 +1,94 @@
-use {super::*, dashmap::DashMap};
+use {
+    super::*,
+    backon::{ExponentialBuilder, Retryable},
+    dashmap::DashMap,
+};
+
+async fn fetch_for<T, P>(client: Client, base: Url, path: String, parse: P) -> (Url, Result<T>)
+where
+    P: FnOnce(&str) -> Result<T> + Send + 'static,
+    T: Send,
+{
+    let result = async {
+        let url = base.join(&path)?;
+        let body = fetch(&client, url).await?;
+        parse(&body)
+    }
+    .await;
+
+    (base, result)
+}
+
+pub async fn fetch(client: &Client, url: Url) -> Result<String> {
+    let started = Instant::now();
+
+    let backoff = ExponentialBuilder::default()
+        .with_min_delay(Duration::from_millis(111))
+        .with_max_delay(Duration::from_millis(999))
+        .with_jitter()
+        .with_max_times(MAX_ATTEMPTS)
+        .with_total_delay(Some(BUDGET));
+
+    let fetch = || async {
+        let now = Instant::now();
+        let elapsed = now.saturating_duration_since(started);
+        if elapsed >= BUDGET {
+            return Err(anyhow!("deadline exceeded"));
+        }
+        let remaining = BUDGET - elapsed;
+        let this_try = remaining.min(TIMEOUT);
+
+        let body = tokio::time::timeout(this_try, async {
+            let response = client.get(url.clone()).send().await?;
+
+            if let Err(status_err) = response.error_for_status_ref() {
+                let text = response.text().await.unwrap_or_default();
+                return Err(anyhow!("{status_err}: {text}").context(status_err));
+            }
+
+            let text = response.text().await?;
+            Ok::<String, Error>(text)
+        })
+        .await
+        .map_err(|_| anyhow::anyhow!("try timeout"))??;
+
+        Ok::<String, Error>(body)
+    };
+
+    let body = fetch
+        .retry(backoff)
+        .sleep(tokio::time::sleep)
+        .when(|err: &Error| {
+            if let Some(err) = err.downcast_ref::<reqwest::Error>() {
+                if err.is_timeout() || err.is_connect() || err.is_request() || err.is_body() {
+                    return true;
+                }
+
+                if let Some(s) = err.status() {
+                    return s == StatusCode::TOO_MANY_REQUESTS || s.is_server_error();
+                }
+
+                return false;
+            }
+
+            let error = err.to_string();
+            if error.contains("try timeout") {
+                return true;
+            }
+
+            if error.contains("deadline exceeded") {
+                return false;
+            }
+
+            false
+        })
```

Comment:
We need to classify unknown/malformed/empty errors. 

Our aggregator can see benefit and we would see less invalid or missing data, especially if we retry on empty response. I'm only seeing logic here for what seems to be the `HTTP 429 Too Many Requests` and `5XX` errors, refusal/reset and build failure.

For example, we can add detection for and retry when there is an empty response because this is common in load balanced or rate-limited APIs and can indicate overload or race conditions, networking glitches that malform into a zero-length response or transient upstream data unavailability. 

```suggestion

            let error = err.to_string();

            if error_str.contains("empty response") {
                return true;
            }
            
            if error.contains("try timeout") {
                return true;
            }

            if error.contains("deadline exceeded") {
                return false;
            }

            false
        })
```